### PR TITLE
Migrate storage from `beta.mycrypto.com`

### PR DESCRIPTION
**Pull request by vilsbole**
Requires https://github.com/MyCryptoHQ/landing/pull/12 to be merged in Landing.  
Replaces https://github.com/MyCryptoHQ/MyCrypto/pull/3415

## Steps to test:
[To test locally, use `https://localhost:8000` for landing and `https://localhost:3000` for app.]
1. Head to [`landing.mycryptobuilds.com`](https://landing.mycryptobuilds.com).
2. Open the console and test that `document.domain` is correctly set to `mycryptobuilds.com`.
3. Add a `v1.0.0` LS with the proper key eg. `MYC_Storage`.
4. Open this PR [link](https://mycryptobuilds.com/pr/3430) in a new tab.
5. Open the console and test that `document.domain` is the identical to the landing page.
6. Open `DevMode` and Reset your localStorage. (this is irreversible so make sure you have a backup before)
7. Refresh your page. (You should arrive to `/no-accounts`
8. See that the banner appears (Should we do a Modal instead?)
9. Click "Yes".
10. Expect: 
- To be redirected to `/dashboard` with your accounts
- See in devTools that the LS updated. You may need to wait. It will take a max of 30secs to see the change.
- See that the storage in landing is deleted.

## Missing:
- [ ] Correct UX for user
- [ ] Unit tests

### CODE REVIEW

**Comment by GuillaumeRx** [comment]
https://github.com/MyCryptoHQ/landing/pull/12 Merged !

File: src/AppRoutes.tsx [hunk 1]
```diff
@@ -2,9 +2,10 @@ import React, { Suspense } from 'react';
...
 import { Layout, LayoutConfig } from '@features/Layout';
-import { Home, PageNotFound, ScreenLockProvider, DrawerProvider } from '@features';
-import { ScrollToTop, useScreenSize } from '@utils';
+import { Home, PageNotFound, ScreenLockProvider, DrawerProvider, MigrateLS } from '@features';
+import { ScrollToTop, useScreenSize, withContext } from '@utils';
 import { useFeatureFlags } from '@services';
+import { StoreContext, SettingsContext } from '@services/Store';
 import { ROUTE_PATHS } from '@config/routePaths';
 import {
   PageVisitsAnalytics,
```
Comment (GuillaumeRx) [inline src/AppRoutes.tsx:11]: Import order

File: src/components/IFrame.test.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+
+import { TURL } from '@types';
+import { default as IFrame } from './IFrame';
+import { simpleRender, screen, fireEvent } from 'test-utils';
```
Comment (GuillaumeRx) [inline src/components/IFrame.test.tsx:5]: Import order

File: src/features/MigrateLS.tsx [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+  isEmpty: boolean;
+  importStorage(ls: LocalStorage): boolean;
+}) => {
+  const [iframeRef, setIframeRef] = useState<HTMLIFrameElement>();
+  const [prevStorage, setPrevStorage] = useState<LocalStorage>();
+  const [error, setError] = useState(false);
+  const src = useMemo(() => getIFrameSrc(getHostName(window)!), []);
+
+  useEffect(() => {
+    console.debug('PrevStorage', prevStorage);
```
Comment (GuillaumeRx) [inline src/features/MigrateLS.tsx:42]: Do we want to keep this ?

File: src/AppRoutes.tsx (hunk 1 above)
Comment (vilsbole) [reply src/AppRoutes.tsx:11]: I made a few changes.
IMHO: `@services` and `@services/Store` belong in the same import category.

File: src/features/MigrateLS.tsx (hunk 3 above)
Comment (vilsbole) [reply src/features/MigrateLS.tsx]: Nope. It has been updated with better error handling.

File: src/components/IFrame.test.tsx (hunk 2 above)
Comment (vilsbole) [reply src/components/IFrame.test.tsx:5]: Done

File: src/AppRoutes.tsx (hunk 1 above)
Comment (vilsbole) [reply src/AppRoutes.tsx:11]: I made some minor changes.
IMHO `@services/Store` should be in the same import block as `@services`.

File: src/features/MigrateLS/MigrateLS.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  text-align: center;
+`;
+
+/**
+ *  Use an IFrame to check if the old url contains a localstorage that we need to import.
+ *  Validate the schema, and prompt the user to import their previous settings.
+ *  If user accepts and migratation succeds:
+ *    - destroy the previous storage so they aren't prompted again.
+ *    - let `NO_ACCOUNTS` handle the redirect, after `Store` is hydrated.
+ *  If user cacels:
```
Comment (FrederikBolding) [inline src/features/MigrateLS/MigrateLS.tsx:35]: typo

File: src/features/MigrateLS/MigrateLS.tsx [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  text-align: center;
+`;
+
+/**
+ *  Use an IFrame to check if the old url contains a localstorage that we need to import.
+ *  Validate the schema, and prompt the user to import their previous settings.
+ *  If user accepts and migratation succeds:
```
Comment (FrederikBolding) [inline src/features/MigrateLS/MigrateLS.tsx:32]: typo

File: src/features/MigrateLS/MigrateLS.test.tsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+};
+
+describe('MigrateLS', () => {
+  test('it is empty by default', () => {
+    const props = { ...defaultProps };
+    getComponent(props);
+    expect(screen.queryByTestId('iframe')).not.toBeInTheDocument();
+  });
+
+  // test('it places a hidden iframe when the user had no custom settings', () => {
```
Comment (FrederikBolding) [inline src/features/MigrateLS/MigrateLS.test.tsx:24]: Finish tests 😄

File: src/features/MigrateLS/MigrateLS.tsx [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  const handleConfirm = () => {
+    downloadAndDestroy();
+  };
+  const handleAbort = () => abortCancel();
+
+  const UI_STATES: Record<UIStates, JSX.Element> = {
+    default: <></>,
+    'migrate-prompt': (
+      <SBanner>
+        We found your previous settings from beta.mycrypto.com. Would you like to import them?
```
Comment (FrederikBolding) [inline src/features/MigrateLS/MigrateLS.tsx:89]: Use translation strings for all of this.

File: src/features/MigrateLS/reducer.test.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+  it('MIGRATE_SUCCESS', () => {
+    const action = {
+      type: MigrateLSReducer.actionTypes.MIGRATE_SUCCESS
+    };
+    const state = dispatch(action)(defaultState);
+    expect(state.isLoading).toEqual(false);
+    expect(state.canDestroy).toEqual(true);
+    expect(state.uiState).toEqual('migrate-success');
+  });
+});
```
Comment (FrederikBolding) [inline src/features/MigrateLS/reducer.test.ts:43]: Add tests for all actionTypes.

File: src/services/Store/StoreProvider.tsx [hunk 9]
```diff
@@ -353,6 +354,14 @@ export const StoreProvider: React.FC = ({ children }) => {
     get defaultAccount() {
       return sortByLabel(state.accounts)[0];
     },
+    /**
+     * Check if the user has already added an account to our persistence layer.
+     */
+    get isDefault() {
+      return (
+        (!state.accounts || isVoid(state.accounts)) && (!isVoid(contacts) || contacts.length < 1)
```
Comment (FrederikBolding) [inline src/services/Store/StoreProvider.tsx:362]: What if they have changed other settings than just adding an account?

File: src/utils/getRootDomain.spec.ts [hunk 10]
```diff
@@ -0,0 +1,25 @@
+import { getRootDomain } from './getRootDomain';
+
+test('it handles undefiend', () => {
```
Comment (FrederikBolding) [inline src/utils/getRootDomain.spec.ts:3]: typo

File: webpack_config/common.js [hunk 11]
```diff
@@ -237,8 +237,8 @@ module.exports = {
       metaCsp: IS_DEV
         ? ''
         : `default-src 'none'; script-src 'self'; worker-src 'self' blob:; child-src 'self'; style-src 'self' 'unsafe-inline'; manifest-src 'self'; font-src 'self'; img-src 'self' data: https://mycryptoapi.com/api/v1/images/; connect-src *${
-        IS_ELECTRON ? ' eth-enclave:' : ''
-        }; frame-src 'self' https://connect.trezor.io;`
+            IS_ELECTRON ? ' eth-enclave:' : ''
+          }; frame-src 'self' https://connect.trezor.io https://landing.mycryptobuilds.com;`
```
Comment (FrederikBolding) [inline webpack_config/common.js:241]: Do we need to add `mycrypto.com` then?

File: src/features/MigrateLS/actions.ts [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+  // Start loading spinner
+  dispatch({ type: Reducer.actionTypes.MIGRATE_REQUEST });
+  try {
+    const success = importFn(storage);
+    if (success) {
+      dispatch({ type: Reducer.actionTypes.MIGRATE_SUCCESS });
+    } else {
+      throw new Error(`[MYC-Migrate] Import failed`);
+    }
+    // destroyStorage(state.iframeRef);
```
Comment (FrederikBolding) [inline src/features/MigrateLS/actions.ts:44]: Remove if not needed.

**Review by FrederikBolding — CHANGES_REQUESTED** [review]
Requested changes.

Current bugs from testing:
- Not downloading correct file
- Not deleting storage on landing after migration

**Comment by FrederikBolding** [comment]
Needs rebase too!

**Comment by vilsbole** [comment]
> Tested and the happy path works great. However: if there is nothing in the landing LS, and no accounts in mycryptobuilds - it shows this in the banner:
> image

Solved

File: src/features/MigrateLS/actions.ts (hunk 12 above)
Comment (vilsbole) [reply src/features/MigrateLS/actions.ts:44]: Done

File: webpack_config/common.js (hunk 11 above)
Comment (vilsbole) [reply webpack_config/common.js:241]: Added `beta`. We should not need the feature when we reach `mycrypto.com`

File: src/utils/getRootDomain.spec.ts (hunk 10 above)
Comment (vilsbole) [reply src/utils/getRootDomain.spec.ts:3]: Done

File: src/features/MigrateLS/MigrateLS.tsx (hunk 7 above)
Comment (vilsbole) [reply src/features/MigrateLS/MigrateLS.tsx:89]: This feature will be deleted before we get translations.
I added a card so we remember to sunset it.
https://app.clubhouse.io/mycrypto/story/6687/sunset-migratels

File: src/features/MigrateLS/MigrateLS.tsx (hunk 4 above)
Comment (vilsbole) [reply src/features/MigrateLS/MigrateLS.tsx:35]: Done

File: src/features/MigrateLS/MigrateLS.tsx (hunk 5 above)
Comment (vilsbole) [reply src/features/MigrateLS/MigrateLS.tsx:32]: Done

**Review by GuillaumeRx — APPROVED** [review]
Code looks good ! Well done 🧙‍♂️

File: src/features/MigrateLS/reducer.test.ts (hunk 8 above)
Comment (vilsbole) [reply src/features/MigrateLS/reducer.test.ts:117]: Done

File: src/features/MigrateLS/MigrateLS.test.tsx (hunk 6 above)
Comment (vilsbole) [reply src/features/MigrateLS/MigrateLS.test.tsx]: Removed commented tests

**Pull request merged**